### PR TITLE
Graphemic lexicon updates / fixes in babel/s5d recipe and hub4_spanish recipe

### PR DESCRIPTION
**Pull request by m-wiesner**
This updates the graphemic lexicon scripts so that they are a little easier to use. Changes include:

- minor differences in handling of special characters
- separate scripts for morfessor training and segmentation in babel s5d recipe
- added option in make_unicode_lexicon.py [--apply-map] for adding lexicon entries for new words
- udpates graphemic lexicon documentation
- adds egs/hub4_spanish/local/expand_graphemic_lexicon.sh which creates the dict and lang 
  directories from a new list of words
- fixes egs/hub4_spanish/local/prepare_lexicon.sh to use updated graphemic lexicon scripts.

### CODE REVIEW

**Comment by danpovey** [comment]
@jtrmal, I'll wait for you to sign off on this before checking it further.
I'm not sure that I like the all-caps filenames.  is that really the first,
most important that we want people entering the babel/s5c directory to see?

**Comment by m-wiesner** [comment]
I agree, but I figured I'd just kept the file names that were already there.

**Comment by danpovey** [comment]
oh I didn't notice they were preexisting.

**Comment by m-wiesner** [comment]
We probably should change it though. We'll see what Yenda has to say.

File: egs/babel/s5d/local/lexicon/apply_morphs.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+. ./utils/parse_options.sh
+
+command -v morfessor-segment >/dev/null 2>&1 || { echo >&2 "Morfessor seems to either not be installed or not on path."; exit 1; }
+
+if [ $# -ne 3 ]; then
+  echo "Usage: ./local/apply_morphs.sh [opts] <morphs_model_dir> <wordlist> <_word2baseform>"
+  echo "  --encoding <encoding_type> # lexicon encoding"
+  echo "  --max-len <int> # maximum allowed morpheme length"  
+  exit 1
```
Comment (jtrmal) [inline egs/babel/s5d/local/lexicon/apply_morphs.sh]: generally speaking, when printing error mesages, the echos should be 
```
echo >&2 "something"
```
typically, we are not extremely careful about it but I think we perhaps should be.

File: egs/babel/s5d/local/lexicon/train_morphs.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+encoding="utf_8"
+
+. ./utils/parse_options.sh
+
+command -v morfessor-train >/dev/null 2>&1 || { echo >&2 "Morfessor seems to either not be installed or not on path."; exit 1; }
+
+if [ $# -ne 2 ]; then
+  echo "Usage: ./local/train_morphs.sh [opts] <path_to_wordcounts> <morphs_directory>" 
+  echo "  --encoding <encoding_type> # lexicon encoding" 
+  exit 1
```
Comment (jtrmal) [inline egs/babel/s5d/local/lexicon/train_morphs.sh]: the same about echos

File: egs/babel/s5d/run-1-main-unicode.sh [hunk 3]
```diff
@@ -96,6 +97,7 @@ if [[ ! -f data/lang/L.fst || data/lang/L.fst -ot $lexicon ]]; then
     data/local $oovSymbol data/local/tmp.lang data/lang
 fi
 
+exit
```
Comment (jtrmal) [inline egs/babel/s5d/run-1-main-unicode.sh]: this is probably mistake?

File: egs/hub4_spanish/s5/local/expand_graphemic_lexicon.sh [hunk 4]
```diff
@@ -0,0 +1,64 @@
+#! /bin/bash
+
+unk="<unk>"
+
+. ./utils/parse_options.sh
+
+if [ $# -eq 0 ]; then
```
Comment (jtrmal) [inline egs/hub4_spanish/s5/local/expand_graphemic_lexicon.sh]: `$# -ne 6`

File: egs/hub4_spanish/s5/local/prepare_unicode_dict.py [hunk 5]
```diff
@@ -3,11 +3,19 @@
...
 # Apache 2.0
 
-# ======= Prepare data/local directory for babel data with unicode tags ======
-# This script creates all files in the data/local directory needed to generate
-# the kaldi-lang directory. It does not generate the lexicon itself -- for that
-# see the script lexicon/make_unicode_lexicon.py. It expects 
+# ======= Prepare dictionary directory (e.g. data/local) from lexicon.txt =====
+# This script creates all files in the dictionary directory.
 #
+# This script basically takes the place of the prepare_lexicon.pl script for
```
Comment (jtrmal) [inline egs/hub4_spanish/s5/local/prepare_unicode_dict.py]: I deleted this intentionally because no-one cares how is it done in babel.
I'd say put the original text back and tweak it to taste (list the files to be created if you want). Keep the reference to lexicon/make_unicode_lexicon.py

File: egs/babel/s5d/local/prepare_unicode_lexicon.py [hunk 6]
```diff
@@ -3,20 +3,19 @@
...
-# 2. nonsilence_phones.txt
-# 3. silence_phones.txt
-# 4. optional_silence.txt
-# 5. extra_questions.txt
+# 1. nonsilence_phones.txt
+# 2. silence_phones.txt
+# 3. optional_silence.txt
+# 4. extra_questions.txt
+#
+# It requires only a valid kaldi format lexicon (lexicon.txt) as input
```
Comment (jtrmal) [inline egs/babel/s5d/local/prepare_unicode_lexicon.py]: add the reference to local/lexicon/make_unicode_lexicon.py

File: egs/babel/s5d/local/prepare_unicode_lexicon.py [hunk 7]
```diff
@@ -25,173 +24,170 @@
 import os
 import argparse
 
-SKIP = ("", "''", "<", ">", "#")
-
 
-# Extract a sorted set of distinct unicode graphemes from the lexicon
-def extract_graphemes(table):
+# Extract a sorted set of distinct phonemes from the lexicon
```
Comment (jtrmal) [inline egs/babel/s5d/local/prepare_unicode_lexicon.py]: when I'm looking at the changes, perhaps important part for a person reading those script would be to understand the distinction between graphemes and phonemes -- can you please document that? Another thing to consider if we want the acoustic units phonemes (but I think that would be ok, if it's documented what that means).

File: egs/babel/s5d/run-1-main-unicode.sh (hunk 3 above)
Comment (m-wiesner) [inline egs/babel/s5d/run-1-main-unicode.sh]: Yeah. Good catch. I was testing the lexicon portion so I added an exit that I forgot to get rid of.

**Comment by m-wiesner** [comment]
Alright. I believe I've committed all the changes you mentioned.

**Comment by jtrmal** [comment]
Switch to "Files changed" and try to review all my comments that are still there, please.

**Comment by jtrmal** [comment]
Scratch that -- as far as I can see, you indeed addressed everything. Sorry. I'll go through it one more time.

File: egs/babel/s5d/local/lexicon/make_word_list.py [hunk 8]
```diff
@@ -59,8 +58,8 @@ def process_transcripts(transcripts_dir, transcripts_list):
 
 def main():
     if len(sys.argv[1:]) == 0:
```
Comment (jtrmal) [inline egs/babel/s5d/local/lexicon/make_word_list.py]: I wonder if you could write len(sys.argv) == 1:

File: egs/babel/s5d/unicode_lexicon_examples.sh [hunk 9]
```diff
@@ -0,0 +1,112 @@
+#!/bin/bash
```
Comment (jtrmal) [inline egs/babel/s5d/unicode_lexicon_examples.sh]: I'm not sure about the purpose of this file? Could you document it's purpose?

File: egs/hub4_spanish/s5/local/expand_graphemic_lexicon.sh [hunk 10]
```diff
@@ -0,0 +1,64 @@
+#! /bin/bash
```
Comment (jtrmal) [inline egs/hub4_spanish/s5/local/expand_graphemic_lexicon.sh]: is this file really needed in the recipe? Or is it just something you wrote for Chunxi?

Comment (m-wiesner) [inline egs/hub4_spanish/s5/local/expand_graphemic_lexicon.sh]: I guess not. It's just there to properly generate a new lexicon from extra text when you have already trained the acoustic models, which was what Chunxi was doing. Should I get rid of it?

File: egs/babel/s5d/unicode_lexicon_examples.sh (hunk 9 above)
Comment (m-wiesner) [inline egs/babel/s5d/unicode_lexicon_examples.sh]: It was meant as part of the readme_unicode file, and it seems I forgot to merge those two files. Is it ok if I put these examples in that file? I wanted to put an example of how to add entries to the lexicon from new external text that you get at a later date without having to retrain the acoustic models, but otherwise, most of these examples are in run-1-main-unicode.sh

**Comment by jtrmal** [comment]
I'd suggest to add some heavily simplified example you feel it might be
useful into the readme but remove this file. Those examples can be pseudo
code, just to show the process, not line-by-line copy and paste code..
Y.

**Comment by jtrmal** [comment]
Yes, please remove it. Y.

**Comment by m-wiesner** [comment]
I removed both files, and just modified readme_unicode a little to incorporate the one part of the unicode_lexicon_examples.sh that I wanted to mention.

**Comment by jtrmal** [comment]
OK, did you run the hub4_spanish just to see it trains and the results are reasonable? If not, can you please do it?

**Comment by jtrmal** [comment]
in any case, update the results file

**Comment by m-wiesner** [comment]
Ok. tri5 is finished. I just ran decode_fmllr_extra.sh and I'm reporting whatever scoring that did.

Previous: 22.92%
My run: 22.53%    When I look at the same LM weight as the previous number it's 22.74%

So it seems to checks out. I'm running the chain tdnn. When you say update the results file, did you mean all the numbers reported in there including the sgmm systems?

**Comment by jtrmal** [comment]
Yes, please update (overwrite) all the results
Y.

**Comment by jtrmal** [comment]
@danpovey  this is ready for merge, I believe
